pkg/reconciler/taskrun/dynamic.go

```diff
@@ -119,33 +117,64 @@ func (r DynamicResolver) Allocate(taskRun *ReconcileTaskRun, ctx context.Context
 			r.eventRecorder.Event(tr, "Normal", "Provisioning", message)
 			err = launchProvisioningTask(taskRun, ctx, tr, secretName, r.sshSecret, address, r.CloudProvider.SshUser(), r.platform, r.sudoCommands)
 			if err != nil {
-				//ugh, try and unassign
+				//Try to delete the instance and unassign it from the TaskRun
 				err := r.CloudProvider.TerminateInstance(taskRun.client, ctx, cloud.InstanceIdentifier(tr.Annotations[CloudInstanceId]))
 				if err != nil {
-					log.Error(err, "Failed to terminate instance")
+					message := fmt.Sprintf("failed to terminate %s instance for %s", r.instanceTag, tr.Name)
+					log.Error(err, message)
 				}
 
 				delete(tr.Labels, AssignedHost)
 				delete(tr.Annotations, CloudInstanceId)
 				delete(tr.Annotations, CloudDynamicPlatform)
 				updateErr := taskRun.client.Update(ctx, tr)
 				if updateErr != nil {
-					log.Error(updateErr, "Could not unassign task after provisioning failure")
+					log.Error(updateErr, "failed to unassign instance from task after provisioning failure")
 					return reconcile.Result{}, err
 				} else {
-					log.Error(err, "Failed to provision cloud host")
+					log.Error(err, "failed to provision cloud host")
 					return reconcile.Result{}, err
 
 				}
 			}
 			return reconcile.Result{}, nil
-		} else {
-			//we are waiting for the instance to come up
-			//so just requeue
+		} else { // A transient error (that wasn't returned) occurred when fetching the IP address for the VM
+			state, err := r.CloudProvider.GetState(taskRun.client, ctx, cloud.InstanceIdentifier(tr.Annotations[CloudInstanceId]))
+			//An error occurred while getting the VM state or it's in a bad state
+			if state == "FAILED" || err != nil {
+				if err != nil {
+					log.Error(
+						err,
+						"failed to get the state for instance",
+						"instanceId", cloud.InstanceIdentifier(tr.Annotations[CloudInstanceId]),
+					)
+				} else { //VM is in a failed state and the task needs to be requeued
+					log.Info("VM instance is in a failed state; will attempt to terminate, unassign from task, and re-queue task")
+				}
+				//Try to delete the instance and unassign it from the TaskRun
+				terr := r.CloudProvider.TerminateInstance(taskRun.client, ctx, cloud.InstanceIdentifier(tr.Annotations[CloudInstanceId]))
```

Comment:
maybe desouple into a function, because very similar code already present above. non insist though, leaving up 2 you.